Simplify methodtable

davidwrighton opened the pull request:
In particular, simplify access to fields of the `MethodTable` structure

Historically, this structure had a great many optional fields in it with several mechanisms for identifying them. As we've been working on CoreCLR, we've eliminated many of these excess fields, to the point where we likely have a final list of all expandable fields associated with the `MethodTable`, and we have space to make access to fields somewhat simpler.

Notably, this is an effort to reduce the number of memory reads and calculations necessary to find the various parts of a `MethodTable`. With this change, I believe that all data directly associated the MethodTable struct can be accessed with no more than 2 pointer chasing memory accesses and possibly reading 1 set of flags. This should simplify much of the codegen around this structure, although performance is likely unaffected, as most of the slow paths were very well optimized by native PGO technologies.

davidwrighton committed:
Simplify MethodTable structure by making the GetModule() lookup simple and the GetLoaderModule path somewhat more complex
- GetModule is more commonly used
- GetLoaderModule is still pretty cheap. It just has 1 extra memory indirection.
- This change adds 1 pointer to the size of all allocated types, which is a bit unfortunate, but it isn't that expensive.

Adjust LoaderHeap handling to pass around LoaderAllocators instead, this is useful for passing through a LoaderAllocator to the CreateMinimalMethodTable code.

davidwrighton committed:
Move NonVirtualSlots table to after the MethodTableWriteableData
- This reduces the complexity of looking it up (it is now straightforward code)
- Also adjust WFLAGS2 so that there are only 8 bits used in it

davidwrighton committed:
Move token to be handled entirely in MethodTable struct directly instead of overflow member

davidwrighton committed:
Move the dispatch map out of being pointed at by a multipurpose slot, into following the MethodTableWriteableData

davidwrighton committed:
Move GenericsStaticsInfo to prepend in front of the MethodTableWriteableData at a fixed offset
- Access to this thing will no longer require computing the number of virtuals
- By being located at a fixed offset, it becomes simpler to debug manually should that be necessary
- Performance of this should be nicely maintained

davidwrighton committed:
Arrays no longer need to set HasPerInstInfo as that flag doesn't control multipurpose slot allocation behavior anymore

davidwrighton committed:
Remove hasInterfaceMap enum flag

davidwrighton committed:
Remove last vestiges of multipurpose slots

davidwrighton committed:
Fix various typos and such so that it builds

davidwrighton committed:
Fix another detail preventing building

davidwrighton committed:
- Fix oops in GetGenericsStaticInfo
- Make DAC a bit more robust against failure
- Fix issue where now that the DispatchMap is being constructed before the MethodTable, we cannot assert if a MethodDesc is virtual

davidwrighton committed:
Fix issue where Array MethodTables didn't have the expected behavior in HasSameTypeDef

davidwrighton committed:
Merge branch 'main' of github.com:dotnet/runtime into simplify_methodtable

davidwrighton committed:
Fix the assert to be correct

davidwrighton committed:
Merge branch 'main' into simplify_methodtable

jkotas commented on src/coreclr/vm/methodtable.cpp:
```diff
@@ -624,8 +516,13 @@ BOOL MethodTable::HasSameTypeDefAs(MethodTable *pMT)
+    // match even if the MethodTable should return a match. So compare
+    // CanonicalMethodTable instead.
+    if (IsArray())
+        return GetCanonicalMethodTable() == pMT->GetCanonicalMethodTable();
```
This looks like a bug in the existing implementation. It does not make sense for `object[]` and `string[]` to have same RID, but `object[]` and `int[]` to not have the same RID. I think this can be simplified to:

```
// optimize for the negative case where we expect RID mismatch
unsigned rid = GetTypeDefRid();
if (rid != pMT->GetTypeDefRid())
    return FALSE;

// Types without RIDs are unrelated to each other. This case is taken for arrays. 
if (rid == 0)
    return FALSE;

return (GetModule() == pMT->GetModule());
```

davidwrighton committed:
Handle the array case of HasSameTypeDefAs consistently
- Array MethodTables now return FALSE for HasSameTypeDefAs
- This required a tweak in CEEInfo::embedGenericHandle to make it behave consistently with previous behavior

davidwrighton committed:
Merge branch 'simplify_methodtable' of github.com:davidwrighton/runtime into simplify_methodtable

jkotas commented on src/coreclr/vm/methodtable.h:
```diff
@@ -296,31 +300,39 @@ struct MethodTableWriteableData
+    };
+    
+
+    PTR_Module m_pLoaderModule;
```
Is there a reason for `m_pLoaderModule` to be here and not in the MethodTable?

davidwrighton replied:
No strong reason, I was trying to avoid growing the `MethodTable` structure, as growth has a (very) small impact on v-table slots due to the limited range of a mov instruction on x64, but its not a critical issue.

Effectively, there is a small size optimization to be had in generated code for virtual function access where if the offset of the vtable chunk index is less than 256 bytes from the start of the `MethodTable`, we can use a 3 byte instruction instead of a 6 byte instruction. It's not significant for throughput perf, but historically I remember seeing some not completely negligible impact on code size for ngen images when adjusting the size of `MethodTable`. However, of course, ngen is no longer a thing, so perhaps I shouldn't worry about that, and looking at the data, I see that it would only matter for virtual table slots 25-32 if we moved two pointer sized values down from `MethodTableWriteableData` to `MethodTable`, which probably isn't significant at all. (Possibly my memories are from the *really* before times, when we didn't have vtable chunks, and this change would have effected vtable slots 7 and 8, which are fairly common in code bases)

From what I can see the `m_pLoaderModule`'s only real perf sensitive use is as part of statics lookup, and I'm planning to make that simpler in a followon checkin, which is why I didn't keep it at the `MethodTable` level, but you are correct, it probably would be fine as a field of MethodTable (along with the `m_hExposedClassObject` field which also has most of the same properties) I'd love to hear if anyone else has an opinion here.

jkotas replied:
I think that the state in this PR is fine. We can fine-tune it in follow up changes.

jkotas commented on src/coreclr/vm/methodtable.h:
```diff
@@ -393,10 +417,30 @@ struct MethodTableWriteableData
+    {
+        return dac_cast<PTR_GenericsStaticsInfo>(dac_cast<TADDR>(pWriteableData) - sizeof(GenericsStaticsInfo));
+    }
+};  // struct MethodTableWriteableData
```
Should `MethodTableWriteableData` be renamed to something that better captures its current purpose? Like `MethodTableAuxiliary` data?

davidwrighton replied:
Ah, great, that's such a nice name. I didn't have a good one in mind, so I wasn't changing it, but MethodTableAuxiliaryData is a distinct improvement, especially since the writeable concept doesn't make any sense anymore.

jkotas reviewed (approved):
LGTM. Thank you!

davidwrighton committed:
Rename all non methodtable WriteableData things to WritableData

davidwrighton committed:
Rename MethodTableWriteableData to MethodTableAuxiliaryData

davidwrighton committed:
Revert "Rename all non methodtable WriteableData things to WritableData"

This reverts commit fb172962a9264e647e1fb897e6d07849eb3cd243.

AaronRobinsonMSFT commented on src/coreclr/vm/methodtablebuilder.cpp:
```diff
@@ -10324,16 +10321,39 @@ MethodTable * MethodTableBuilder::AllocateNewMT(
+        pDispatchMap = new (pAllocatedSpaceAfterMethodTableAuxiliaryData) DispatchMap(pbDispatchMapTemp, cbDispatchMapTemp);
+
+#ifdef LOGGING
+        if (dispatchMapAllocationSize.IsOverflow())
+            ThrowHR(E_FAIL);
```
This seems odd to be behind the `LOGGING` define and it is a bit far away from where `dispatchMapAllocationSize` is defined. I would expect this up a bit.

davidwrighton replied:
So, `dispatchMapAllocationSize` is a `S_SIZE_T`, and those need to be Overflow checked before they can be read. `AllocateAuxiliaryData` does that internally, but its overflow check does not escape back to this function, and so we need an additional check for overflow here. (Note: this check should never fail, as its already being used above in `AllocateAuxiliaryData`

davidwrighton replied:
Of course, this doesn't even need to be an `S_SIZE_T` ...so I can fix that.

AaronRobinsonMSFT commented on src/coreclr/vm/methodtablebuilder.cpp:
```diff
@@ -10324,16 +10321,39 @@ MethodTable * MethodTableBuilder::AllocateNewMT(
+        pMT->SetFlag(MethodTable::enum_flag_HasPerInstInfo);
+    }
+
+    if (bmtVT->pDispatchMapBuilder->Count() > 0)
+        pMT->SetFlag(MethodTable::enum_flag_HasDispatchMapSlot);
```
We are checking `pDispatchMapBuilder->Count() > 0` twice and very close to each other. Can they be collapsed?

davidwrighton replied:
Hmm... Probably. I'll give it a try.

davidwrighton replied:
Yep Looks like a fine idea. I've collapsed them.

AaronRobinsonMSFT commented on src/coreclr/vm/methodtablebuilder.cpp:
```diff
@@ -10324,16 +10321,39 @@ MethodTable * MethodTableBuilder::AllocateNewMT(
-    pMT->SetWriteableData(pMTWriteableData);
+    pMT->AllocateAuxiliaryData(pAllocator, pLoaderModule, pamTracker, fHasGenericsStaticsInfo, static_cast<WORD>(dwNonVirtualSlots), dispatchMapAllocationSize);
 
     // This also disables IBC logging until the type is sufficiently initialized so
     // it needs to be done early
```
Is this comment still relevant?

elinor-fung commented on src/coreclr/vm/methodtable.cpp:
```diff
@@ -809,11 +703,38 @@ MethodDesc *MethodTable::GetMethodDescForComInterfaceMethod(MethodDesc *pItfMD,
+
+    size_t prependedAllocationSpace = 0;
+
+    prependedAllocationSpace = nonVirtualSlots * sizeof(TADDR);
```
Just initialize `prependedAllocationSpace` to this instead of 0?

davidwrighton committed:
Code review tweaks

AaronRobinsonMSFT commented:
The only other thing I'd ask is if there is anything that needs to be updated in the BoTR. I did a quick check and nothing jumped out, but it would be helpful if you could do a quick check.

davidwrighton commented:
@AaronRobinsonMSFT  I've taken a quick look at the BoTR. All these details have apparently fallen under the category of trivial details not mentioned in the docs.
